File: apps/api/src/frw_api/services/audit.py
from __future__ import annotations

import hashlib
import json
from typing import Any

from sqlalchemy import text
from sqlalchemy.orm import Session

from frw_api.auth.security import CurrentUser


def stable_hash(value: Any) -> str:
    payload = json.dumps(value, sort_keys=True, default=str, separators=(",", ":"))
    return "sha256:" + hashlib.sha256(payload.encode()).hexdigest()


def audit(
    db: Session,
    *,
    user: CurrentUser | None,
    action: str,
    target_table: str | None = None,
    target_pk: str | None = None,
    before: Any = None,
    after: Any = None,
    request_id: str | None = None,
) -> None:
    db.execute(
        text(
            """
            insert into audit_log(
              actor_user_id, actor_role, action, target_table, target_pk,
              before_hash, after_hash, request_id
            )
            values (
              :actor_user_id, :actor_role, :action, :target_table, :target_pk,
              :before_hash, :after_hash, :request_id
            )
            """
        ),
        {
            "actor_user_id": user.id if user else None,
            "actor_role": user.role if user else None,
            "action": action,
            "target_table": target_table,
            "target_pk": target_pk,
            "before_hash": stable_hash(before) if before is not None else None,
            "after_hash": stable_hash(after) if after is not None else None,
            "request_id": request_id,
        },
    )

File: apps/backend_elixir/lib/stonks_backend/audit.ex
defmodule StonksBackend.Audit do
  @moduledoc "Audit log compatibility writer."

  alias StonksBackend.Sql

  require Logger

  def write(action, opts \\ []) do
    user = Keyword.get(opts, :user)
    target_table = Keyword.get(opts, :target_table)
    target_pk = Keyword.get(opts, :target_pk)
    after_payload = Keyword.get(opts, :after, %{})

    Sql.execute(
      """
      insert into audit_log(actor_user_id, actor_role, action, target_table, target_pk, after_json)
      values ($1, $2, $3, $4, $5, $6::text::jsonb)
      """,
      [
        user && user[:id],
        user && user[:role],
        action,
        target_table,
        target_pk,
        Jason.encode!(after_payload || %{})
      ]
    )
  rescue
    _ ->
      Logger.error("Audit log persistence failed action=#{action}")
      {:error, :storage_unavailable}
  end
end
